services/src/security_mode.cpp

```diff
@@ -16,21 +16,121 @@
+
+void timerCallback(os_timer_t timer);
+
+void stopTimer() {
+    if (sTimer) {
+        int r = os_timer_change(sTimer, OS_TIMER_CHANGE_STOP, false /* fromISR */, 0 /* period */, 0xffffffffu /* block */, nullptr /* reserved */);
+        if (r != 0) {
+            LOG_DEBUG(ERROR, "Failed to stop timer"); // Should not happen
+        }
+    }
+}
+
+int startTimer() {
+    if (!sTimer) {
+        int r = os_timer_create(&sTimer, 1000 /* period */, timerCallback, nullptr /* timer_id */, false /* one_shot */, nullptr /* reserved */);
+        if (r != 0) {
+            return SYSTEM_ERROR_NO_MEMORY;
+        }
+    } else {
+        stopTimer();
+    }
+    int r = os_timer_change(sTimer, OS_TIMER_CHANGE_START, false /* fromISR */, 0 /* period */, 0xffffffffu /* block */, nullptr /* reserved */);
+    if (r != 0) {
+        return SYSTEM_ERROR_UNKNOWN; // Should not happen
+    }
+    return 0;
+}
+
+void timerCallback(os_timer_t) {
+    bool stop = false;
+    ATOMIC_BLOCK() {
+        Load_SystemFlags();
+        if (system_flags.security_mode_override_value != 0xff) {
+            if (system_flags.security_mode_override_timeout > 0) {
+                --system_flags.security_mode_override_timeout;
+            }
+            if (!system_flags.security_mode_override_timeout) {
+                system_flags.security_mode_override_value = 0xff;
+            }
+            Save_SystemFlags();
+        }
+        if (system_flags.security_mode_override_value == 0xff) {
+            sCurrentSecurityMode = sNormalSecurityMode;
+            stop = true;
+        }
+    }
+    if (stop) {
+        stopTimer();
+    }
+}
+
+#endif // MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
 
 } // namespace
 
-int security_mode_find_extension(hal_storage_id storageId, uintptr_t start, module_info_security_mode_ext_t* securityModeExt) {
+int security_mode_init() {
+    // Note that this function is called early on boot, before the heap is configured and C++
+    // constructors are called
+    if (!FLASH_VerifyCRC32(FLASH_INTERNAL, module_bootloader.start_address, FLASH_ModuleLength(FLASH_INTERNAL, module_bootloader.start_address))) {
+        return SYSTEM_ERROR_BAD_DATA;
+    }
+    module_info_security_mode_ext_t ext = {};
+    ext.ext.length = sizeof(ext);
+    int r = security_mode_find_module_extension(HAL_STORAGE_ID_INTERNAL_FLASH, module_bootloader.start_address, &ext);
+    // XXX: security_mode_find_module_extension() should ideally not return an error if the bootloader
+    // is not protected
+    if (r < 0 && r != SYSTEM_ERROR_NOT_FOUND) {
+        return r;
+    }
+    if (ext.security_mode == MODULE_INFO_SECURITY_MODE_PROTECTED) {
+        sNormalSecurityMode = ext.security_mode;
+        Load_SystemFlags();
+        if (system_flags.security_mode_override_value != 0xff) {
+            sCurrentSecurityMode = system_flags.security_mode_override_value;
+        } else {
+            sCurrentSecurityMode = sNormalSecurityMode;
+        }
+    }
```

Comment:
I think you need an else to clear the security mode override if the device is not protected, otherwise the following situation can happen:
* device is protected
* device gets unlocked (protected = false, overriden = true)
* open bootloader gets flashed (protected = false, overriden = true)
* protected bootloader gets flashed. At this point, protected should be true and overriden should be false, but because the `security_mode_override_value` wasn't cleared, we get protected = false, overriden = true

Reply:
Good catch, @keeramis! Fixed in https://github.com/particle-iot/device-os/pull/2770/commits/c116ce7e55a9710612c8e5eb07dceea2f21b5496.

Reply:
Ended up refactoring that function so that it would also clear the override if it for whatever reason failed to parse the bootloader extensions.